src/contracts/core/DelegationManager.sol

```diff
@@ -185,46 +185,59 @@ contract DelegationManager is Initializable, OwnableUpgradeable, Pausable, Deleg
     }
 
     /**
-     * @notice Undelegates `staker` from the operator who they are delegated to.
-     * @param staker The account undelegating.
+     * @notice Undelegates the staker from the operator who they are delegated to. Puts the staker into the "undelegation limbo" mode of the EigenPodManager
+     * and queues a withdrawal of all of the staker's shares in the StrategyManager (to the staker), if necessary.
+     * @param staker The account to be undelegated.
+     * @return withdrawalRoot The root of the newly queued withdrawal, if a withdrawal was queued. Otherwise just bytes32(0).
      *
-     * @dev Callable only by the StrategyManager.
-     * @dev Should only ever be called in the event that the `staker` has no active deposits in EigenLayer.
      * @dev Reverts if the `staker` is also an operator, since operators are not allowed to undelegate from themselves.
-     * @dev Does nothing (but should not revert) if the staker is already undelegated.
+     * @dev Reverts if the caller is not the staker, nor the operator who the staker is delegated to, nor the operator's specified "delegationApprover"
+     * @dev Reverts if the `staker` is already undelegated.
      */
-    function undelegate(address staker) external onlyStrategyManagerOrEigenPodManager {
-        require(!isOperator(staker), "DelegationManager.undelegate: operators cannot undelegate from themselves");
+    function undelegate(address staker) external returns (bytes32 withdrawalRoot) {
+        require(isDelegated(staker), "DelegationManager.undelegate: staker must be delegated to undelegate");
         address operator = delegatedTo[staker];
-        // only make storage changes + emit an event if the staker is actively delegated, otherwise do nothing
-        if (operator != address(0)) {
-            emit StakerUndelegated(staker, operator);
-            delegatedTo[staker] = address(0);
+        require(!isOperator(staker), "DelegationManager.undelegate: operators cannot be undelegated");
+        require(staker != address(0), "DelegationManager.undelegate: cannot undelegate zero address");
+        require(
+            msg.sender == staker ||
+            msg.sender == operator ||
+            msg.sender == _operatorDetails[operator].delegationApprover,
+            "DelegationManager.undelegate: caller cannot undelegate staker"
+        );
+        
+        if (!canUndelegate(staker)) {
```

Comment:
Is this just for EigenPod owner (stakers)?

Reply:
This is a result of combining the `undelegate` and `forceUndelegate` functions. See my comment above -- I think renaming the function to `hasActivelyDelegatedShares` is probably the right change to be made here; I can also add a comment to the effect of
```
// remove any shares from the delegation system that the staker currently has delegated, if necessary
```